presto-main/src/test/java/com/facebook/presto/server/security/TestPrestoAuthenticator.java

```diff
@@ -0,0 +1,141 @@
+                ImmutableListMultimap.of(TEST_HEADER, TEST_HEADER_VALID_VALUE + ":" + TEST_USER),
+                TEST_REMOTE_ADDRESS,
+                ImmutableMap.of());
+
+        Optional<Principal> principal = checkAuthentication(prestoAuthenticatorManager.getAuthenticator(), request);
+        assertTrue(principal.isPresent());
+        assertEquals(principal.get().getName(), TEST_USER);
+
+        // Test failed authentication
+        request = new MockHttpServletRequest(
+                ImmutableListMultimap.of(TEST_HEADER, TEST_HEADER_INVALID_VALUE + ":" + TEST_USER),
+                TEST_REMOTE_ADDRESS,
+                ImmutableMap.of());
+
+        principal = checkAuthentication(prestoAuthenticatorManager.getAuthenticator(), request);
+        assertFalse(principal.isPresent());
+    }
+
+    private Optional<Principal> checkAuthentication(PrestoAuthenticator authenticator, HttpServletRequest request)
+    {
+        try {
+            // Converting HttpServletRequest to Map<String, String>
+            Map<String, String> headers = getHeadersMap(request);
+
+            // Passing the headers Map to the authenticator
+            return Optional.of(authenticator.createAuthenticatedPrincipal(headers));
+        }
+        catch (AccessDeniedException e) {
+            return Optional.empty();
+        }
+    }
+
+    private Map<String, String> getHeadersMap(HttpServletRequest request)
+    {
+        Map<String, String> headersMap = new HashMap<>();
+        Enumeration<String> headerNames = request.getHeaderNames();
+        while (headerNames.hasMoreElements()) {
+            String headerName = headerNames.nextElement();
+            String headerValue = request.getHeader(headerName);
+            headersMap.put(headerName, headerValue);
+        }
+        return headersMap;
+    }
+
+    private static class TestingPrestoAuthenticatorFactory
+            implements PrestoAuthenticatorFactory
+    {
+        private final String name;
+        private final String validHeaderValue;
+
+        TestingPrestoAuthenticatorFactory(String name, String validHeaderValue)
+        {
+            this.name = requireNonNull(name, "name is null");
+            this.validHeaderValue = requireNonNull(validHeaderValue, "validHeaderValue is null");
+        }
+
+        @Override
+        public String getName()
+        {
+            return this.name;
+        }
+
+        @Override
+        public PrestoAuthenticator create(Map<String, String> config)
+        {
+            return (httpRequest) -> {
+                // TEST_HEADER will have value of the form PART1:PART2
+                String[] header = httpRequest.get(TEST_HEADER).split(":");
+
+                if (header[0].equals(this.validHeaderValue)) {
+                    return new BasicPrincipal(header[1]);
+                }
+
+                throw new AccessDeniedException("Authentication Failed!");
+            };
```

Comment:
After making the headers map in SPI to `Map<String, List<String>>`, we can refactor this.
```suggestion
            return (headers) -> {
                // TEST_HEADER will have value of the form PART1:PART2
                String[] header = headers.get(TEST_HEADER).get(0).split(":");

                if (header[0].equals(this.validHeaderValue)) {
                    return new BasicPrincipal(header[1]);
                }

                throw new AccessDeniedException("Authentication Failed!");
            };
```